### PR TITLE
Batched weighting in `calc_dcd()`

### PR DESCRIPTION
**Pull request by kazuto1011**
Thank you for sharing your work. In this PR, I modified the weighting part in `calc_dcd()` to a batched manner. Although the memory consumption may increase, the speed is much faster!
This PR only modifies `utils_v2/` because I have an issue with compiling `utils/`.

The attached scripts gave the following results with my machine.

```py
orig: 6.357818365097046 [s] # original
new : 0.5714225769042969 [s] # with this PR
dcd_orig=tensor([0.5886, 0.5782, 0.5761, 0.5773, 0.5795, 0.5833, 0.5820, 0.5802, 0.5821,
        0.5912, 0.5872, 0.5813, 0.5801, 0.5830, 0.5833, 0.5876],
       device='cuda:0')
dcd_new=tensor([0.5886, 0.5782, 0.5761, 0.5773, 0.5795, 0.5833, 0.5820, 0.5802, 0.5821,
        0.5912, 0.5872, 0.5813, 0.5801, 0.5830, 0.5833, 0.5876],
       device='cuda:0')
```

```py
import time

import torch

from utils_v2.model_utils import calc_cd
from utils_v2.model_utils import calc_dcd as calc_dcd_orig


def calc_dcd_new(x, gt, alpha=1000, n_lambda=1, return_raw=False, non_reg=False):
    x = x.float()
    gt = gt.float()
    _, n_x, _ = x.shape
    _, n_gt, _ = gt.shape
    assert x.shape[0] == gt.shape[0]

    if non_reg:
        frac_12 = max(1, n_x / n_gt)
        frac_21 = max(1, n_gt / n_x)
    else:
        frac_12 = n_x / n_gt
        frac_21 = n_gt / n_x
    cd_p, cd_t, dist1, dist2, idx1, idx2 = calc_cd(x, gt, return_raw=True)
    # dist1 (batch_size, n_gt): a gt point finds its nearest neighbour x' in x;
    # idx1  (batch_size, n_gt): the idx of x' \in [0, n_x-1]
    # dist2 and idx2: vice versa
    exp_dist1, exp_dist2 = torch.exp(-dist1 * alpha), torch.exp(-dist2 * alpha)

    # batched!
    count1 = torch.zeros_like(idx2)
    count1.scatter_add_(1, idx1.long(), torch.ones_like(idx1))
    weight1 = count1.gather(1, idx1.long()).float().detach() ** n_lambda
    weight1 = (weight1 + 1e-6) ** (-1) * frac_21
    loss1 = (1 - exp_dist1 * weight1).mean(dim=1)

    # batched!
    count2 = torch.zeros_like(idx1)
    count2.scatter_add_(1, idx2.long(), torch.ones_like(idx2))
    weight2 = count2.gather(1, idx2.long()).float().detach() ** n_lambda
    weight2 = (weight2 + 1e-6) ** (-1) * frac_12
    loss2 = (1 - exp_dist2 * weight2).mean(dim=1)

    loss = (loss1 + loss2) / 2

    res = [loss, cd_p, cd_t]
    if return_raw:
        res.extend([dist1, dist2, idx1, idx2])

    return res


pc_1 = torch.randn([16, 2048, 3]).cuda() * 0.1
pc_2 = torch.randn([16, 2048, 3]).cuda() * 0.1

torch.cuda.synchronize()
start_t = time.time()
for _ in range(1000):
    dcd_orig, _, _ = calc_dcd_orig(pc_1, pc_2)
torch.cuda.synchronize()
print(f"orig: {time.time() - start_t} [s]")

torch.cuda.synchronize()
start_t = time.time()
for _ in range(1000):
    dcd_new, _, _ = calc_dcd_new(pc_1, pc_2)
torch.cuda.synchronize()
print(f"new : {time.time() - start_t} [s]")

print(f"{dcd_orig=}")
print(f"{dcd_new=}")
```

### CODE REVIEW

**Comment by wutong16** [comment]
Sorry for the delay. 

I've checked the loss curve of the training process and it runs correctly.

Thanks very much for your contribution!
